finagle-memcached/src/main/scala/com/twitter/finagle/Memcached.scala

```diff
@@ -22,32 +22,31 @@ import com.twitter.finagle.tracing._
 import com.twitter.finagle.transport.Transport
 import com.twitter.hashing
 import com.twitter.io.Buf
-import com.twitter.util.{Closable, Duration, Future, Monitor}
+import com.twitter.util.{Closable, Duration, Monitor}
 import com.twitter.util.registry.GlobalRegistry
 import scala.collection.mutable
 
-/**
- * Defines a [[Tracer]] that understands the finagle-memcached request type.
- * This is installed as the `ClientTracingFilter` in the default stack.
- */
-private[finagle] object MemcachedTraceInitializer {
-  object Module extends Stack.Module1[param.Tracer, ServiceFactory[Command, Response]] {
-    val role = TraceInitializerFilter.role
-    val description = "Initialize traces for the client and record hits/misses"
-    def make(_tracer: param.Tracer, next: ServiceFactory[Command, Response]) = {
-      val param.Tracer(tracer) = _tracer
-      val filter = new Filter(tracer)
-      filter andThen next
+private[finagle] object MemcachedTracingFilter {
+
+  object Module extends Stack.Module1[param.Label, ServiceFactory[Command, Response]] {
+    val role = ClientTracingFilter.role
+    val description = "Add Memcached client specific annotations to the trace"
+
+    def make(_label: param.Label, next: ServiceFactory[Command, Response]) = {
+      val param.Label(label) = _label
+      val annotations = new AnnotatingTracingFilter[Command, Response](
+        label, Annotation.ClientSend(), Annotation.ClientRecv())
+      annotations andThen TracingFilter andThen next
     }
   }
 
-  class Filter(tracer: Tracer) extends SimpleFilter[Command, Response] {
-    def apply(command: Command, service: Service[Command, Response]): Future[Response] =
-      Trace.letTracerAndNextId(tracer) {
-        val response = service(command)
+  object TracingFilter extends SimpleFilter[Command, Response] {
+    def apply(command: Command, service: Service[Command, Response]) = {
+      val response = service(command)
```

Comment:
Step #3 represents the rest of the client stack and we have no guarantees
that there's no tracing happening there so ideally we'd do the request
level tracing before dispatching the request through the rest of the stack.

I'm okay with leaving it as-is but can you leave a comment that this
implementation assumes no further tracing lower in the stack?

On Fri, Jun 17, 2016 at 1:04 AM, Kristof Adriaenssens <
notifications@github.com> wrote:

> In finagle-memcached/src/main/scala/com/twitter/finagle/Memcached.scala
> <https://github.com/twitter/finagle/pull/516#discussion_r67473223>:
>
> >      }
> >    }
> >
> > -  class Filter(tracer: Tracer) extends SimpleFilter[Command, Response] {
> > -    def apply(command: Command, service: Service[Command, Response]): Future[Response] =
> > -      Trace.letTracerAndNextId(tracer) {
> > -        val response = service(command)
> > +  object TracingFilter extends SimpleFilter[Command, Response] {
> > +    def apply(command: Command, service: Service[Command, Response]) = {
> > +      val response = service(command)
>
> I think we are fine with these changes.
> The order in which tracing events happen for memcached with this commit in
> place:
>
>    1. Initialize span (TraceInitializerFilter part of StackClient)
>    2. Add Client Send annotation (AnnotatingTracingFilter set-up in
>    MemcachedTracingFilter.Module)
>    3. Execute memcached request
>    4. Add cache hit/miss annotations based on memcached response
>    (MemcachedTracingFilter)
>    5. Add Client Received annotation (AnnotatingTracingFilter set-up in
>    MemcachedTracingFilter.Module)
>
> We can only add the cache hit/miss annotations once we got the response.
> The important change is that the memcached annotations are submitted in
> between the cs / cr annotations. The Client Received annotation should be
> the last annotation that gets submitted because it indicates the span is
> complete and ready to send off to back-end.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/twitter/finagle/pull/516/files/67af7342d082c4408f125b8d15cf68a394506b2a#r67473223>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AADkAZf67j9RBSfbjaWWpioBOLc6H9-Dks5qMlUDgaJpZM4I0Dkv>
> .
>